pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -882,30 +885,40 @@ func findNextInitContainerToRun(pod *v1.Pod, podStatus *kubecontainer.PodStatus)
 	}
 
 	// There are no failed containers now.
-	for i := len(pod.Spec.InitContainers) - 1; i >= 0; i-- {
-		container := &pod.Spec.InitContainers[i]
+	for _, container := range pod.Spec.InitContainers {
+		if hasInitialized && !types.IsSidecarContainer(&container) {
+			continue
+		}
+
 		status := podStatus.FindContainerStatusByName(container.Name)
 		if status == nil {
-			continue
+			return nil, &container, false
 		}
 
-		// container is still running, return not done.
-		if status.State == kubecontainer.ContainerStateRunning {
+		switch status.State {
+		case kubecontainer.ContainerStateCreated:
 			return nil, nil, false
-		}
 
-		if status.State == kubecontainer.ContainerStateExited {
-			// all init containers successful
-			if i == (len(pod.Spec.InitContainers) - 1) {
-				return nil, nil, true
+		case kubecontainer.ContainerStateRunning:
+			if !types.IsSidecarContainer(&container) {
+				return nil, nil, false
 			}
 
-			// all containers up to i successful, go to i+1
-			return nil, &pod.Spec.InitContainers[i+1], false
+			if startup, found := m.startupManager.Get(status.ID); found && startup == proberesults.Failure {
+				return nil, nil, false
+			}
+			// sidecar container has started, continue
+
+		case kubecontainer.ContainerStateExited:
+			if types.IsSidecarContainer(&container) {
+				return nil, &container, false
+			}
```

Comment:
```suggestion
			}
			// normal initContainer has exited, continue
```

(t align with explanation of other cases when no return happens)?